### ISSUE TITLE
wheels segfault when on CUDA

### ISSUE THREAD

**Issue by bertsky** [shapely/shapely#1598]
## Expected behavior and actual behavior.

Something is still wrong with the 1.8.5.post1 wheels: When importing from Shapely in a certain way, I later-on get segfaults when running code on the GPU.

(My guess is that somehow the way the post-release was done in https://github.com/shapely/shapely/issues/1567 is unusual and broken.)

## Steps to reproduce the problem.

```
python -c "from shapely.geometry import Polygon; import torch; torch.randn(10).cuda()"
Segmentation fault (core dumped)
```

(This is the minimal example I could come up with. There are lots of real segfaults when doing other CUDA stuff with this version of Shapely.)

It goes away with importing **all** of Shapely. 

It also goes away when installing 1.8.5.post1 from source (either `pip install --no-binary shapely --force-reinstall shapely` or `pip install --force-reinstall git+https://github.com/shapely/shapely@maint-1.8`).

## Operating system

This is Linux with Python 3.8. (I tried different versions of Pytorch and CUDA toolkit – always the same.)

## Shapely version and provenance

1.8.5.post1 wheel as published on PyPI

**Comment by jorisvandenbossche** [comment]
See also https://github.com/shapely/shapely/issues/1435 for a similar report. 
I would personally recommend to start using shapely 2.0 which fixes this problem (it is unclear how to solve this for 1.8.x, although contributions are certainly welcome if you can further investigate).

**Comment by bertsky** [comment]
Well, like I said, doing `import shapely` (all of it) makes the problem disappear, so it should be solvable for 1.8.x IMO.

(I did try `shapely.speedups.disable()` as [suggested](https://github.com/shapely/shapely/issues/1435#issuecomment-1170689840) by @ZhuYQi, but that did not change anything. In contrast, importing pytorch first did avoid the problem, as described for #1435 as well.)

Of course we might migrate to Shapely 2.0, but I am afraid we will run into other issues with that, so I don't want to be forced to updated yet.

**Comment by jorisvandenbossche** [comment]
> Well, like I said, doing `import shapely` (all of it) makes the problem disappear, so it should be solvable for 1.8.x IMO.

You can also use that as a workaround for now? 
Or we can do that in shapely, you mean, as a solution? I assume it is import of shapely.speedups and shapely.vectorized that is the most important (as those are compiled instead of using ctypes)

> It also goes away when installing 1.8.5.post1 from source

Yes, that way you don't get the GEOS that is included in the wheel, and you ensure shapely will always use the GEOS available on your system.

**Comment by jorisvandenbossche** [comment]
> Of course we might migrate to Shapely 2.0, but I am afraid we will run into other issues with that, so I don't want to be forced to updated yet.

I certainly understand you cannot require that right now (certainly given we are still in betal), but I would certainly recommend in already trying it out, to ensure we find those other issues and can fix them.

**Comment by bertsky** [comment]
> > Well, like I said, doing `import shapely` (all of it) makes the problem disappear, so it should be solvable for 1.8.x IMO.
> 
> You can also use that as a workaround for now?

Unfortunately not. We are developing a large [OCR framework](https://github.com/OCR-D/ocrd_all) with lots of independent submodules, many of which use Shapely in some form. It's difficult to enforce a certain import style.

(BTW, `import shapely` of course does not do anything, because it's just the namespace pkg on 1.8.x. But `from shapely import *` works without crashing. So that's the workaround I actually meant.)

> Or we can do that in shapely, you mean, as a solution? I assume it is import of shapely.speedups and shapely.vectorized that is the most important (as those are compiled instead of using ctypes)

Well, since most submodules/subpackages pull `geos`, which already crashes, it must be something loaded **before** that is relevant.

It turns out: the only way other than `from shapely import *` is by importing `_buildcfg` first!

> > It also goes away when installing 1.8.5.post1 from source
> 
> Yes, that way you don't get the GEOS that is included in the wheel, and you ensure shapely will always use the GEOS available on your system.

Yes, I understand that. The bundled libgeos might be the problem already. But see above – seems to be ok if `_buildcfg` is there already.

**Comment by bertsky** [comment]
> I certainly understand you cannot require that right now (certainly given we are still in betal), but I would certainly recommend in already trying it out, to ensure we find those other issues and can fix them.

Yes, the problem is that (in my experience) most errors only happen rarely, as they depend on so much from the data side. As soon as we have our data regression test ready, we'll include prereleases in CI to get early warnings.

**Comment by jorisvandenbossche** [comment]
> (BTW, `import shapely` of course does not do anything, because it's just the namespace pkg on 1.8.x. But `from shapely import *` works without crashing. So that's the workaround I actually meant.)

What I think is happening here is that `from shapely import *` is _also_ not doing anything (for the same reason). So the reason that this "workaround" seems to work, is also because it just not yet imports anything, just like `import shapely` (and in contrast with `import shapely.geometry`).

So the issue is still that in _general_ importing actual shapely functionality before importing torch gives segfaults.

**Comment by stweil** [comment]
With Python 3.11, shapely-1.8.5.post1, torch-2.0.1 and the original example I don't get a crash, but the program hangs (at least 30 seconds). But `python -c "import torch; from shapely.geometry import Polygon; torch.randn(10).cuda()"` works fine. So the hint to import torch before shapely works fine for me.

**Comment by stweil** [comment]
> It also goes away when installing 1.8.5.post1 from source (either `pip install --no-binary shapely --force-reinstall shapely` or `pip install --force-reinstall git+https://github.com/shapely/shapely@maint-1.8`).

@bertsky, these commands don't create site-packages/Shapely.libs which is installed by the normal installation `pip install shapely`. The normal installation seems to work, too, after removing site-packages/Shapely.libs, so that looks like another possible workaround. @jorisvandenbossche, maybe that helps as a hint for fixing this issue?

**Comment by bertsky** [comment]
@stweil
> shapely-2.0.1 works with any order of import statements.

I cannot confirm that. As [explained here](https://github.com/OCR-D/core/pull/1041#discussion_r1165323116) I still saw the same problem there. Maybe it depends on whether wheels are already available for your version / platform?

**Comment by jorisvandenbossche** [comment]
@bertsky do you remember the details of how you tested this? (which OS, python version, how did you install it, etc) 
Because on CI of the PR you linked it seemed to actually work fine without forcing the no-binary install?

**Comment by stweil** [comment]
@bertsky, could you please try whether the workaround with removing site-packages/Shapely.libs works for shapely-2.0.1, too?

I tested on Debian stable with its default Python 3.11.
